Fixes #2, replaces direct GuzzleHttp with PHP League Oauth2 client

tm1000 opened the pull request:
Fixes #2 by adding PHP League Oauth 2.0 Client. This could be a drop in replacement but I had to change one function's description, that being setAccessToken because I am using the objects returned from PHP League instead of pure strings

**_Note:_** The PHP Unit tests will always fail because config.json is always empty and it can't work against Zoho directly. My suggestion here is that the Unit tests shouldn't action against Zoho directly anyways they need to be self contained and only test the functions themselves. Otherwise you have to store working credentials in this library to utilize it.

This also adds a function to get the resource owner.

Example:
```php
$filesystemAdapter = new Local(__DIR__ . '/tests/temp');
$filesystem        = new Filesystem($filesystemAdapter);
$pool = new FilesystemCachePool($filesystem);

$oAuthClient = new OAuthClient(<client>, <secret>);
$oAuthClient->setRefreshToken(<refreshtoken>);
$oAuthClient->useCache($pool);
$oAuthClient->offlineMode();
$oAuthClient->setRegion(\Weble\ZohoClient\Enums\Region::us());
var_dump($oAuthClient->getAccessToken());
$user = $oAuthClient->getResourceOwner();
var_dump($user);
```

Returns
```php
/usr/src/ZohoClient/test.php:20:
string(70) "<accesstoken>"
/usr/src/ZohoClient/test.php:22:
class Weble\ZohoClient\Provider\ZohoUser#26 (1) {
  protected $response =>
  array(5) {
    'First_Name' =>
    string(8) "Zoho API"
    'Email' =>
    string(25) "<email>"
    'Last_Name' =>
    string(4) "User"
    'Display_Name' =>
    string(8) "Zoho API"
    'ZUID' =>
    int(<zuid>)
  }
}
```

tm1000 committed:
Fixes #2, replaces direct GuzzleHttp with PHP League Oauth2 client

tm1000 committed:
require php league

Skullbock commented:
@tm1000 thanks for the PR! is there a reason why guzzle is still a dependency?

As for the tests, they are failing because i din't yet setup the github actions to run with my custom auth.

I'm not sure i agree with 
`My suggestion here is that the Unit tests shouldn't action against Zoho directly anyways they need to be self contained and only test the functions themselves. Otherwise you have to store working credentials in this library to utilize it.`

In my (humble) opinion, a library which only job is to interact with a 3rd party api, SHOULD test against the real endpoints. Otherwise the tests just check the internal workings, and although that's fine, i would rather have a test that does that AND checks that the library interacts with the API correctly. More so with zoho, where they change / break stuff often 👍 

I'll see if i can re-enable the automated testing against my dev instance of zoho crm

Cheers

tm1000 commented:
@Skullbock My only concern was with you exposing credentials you dont want to expose. However if github actions allow you to add credentials without showing the rest of us then it's a moot point :-)

tm1000 commented:
Oh and Guzzle is still a dep because I was catching a guzzle exception of ClientException in ONE place. We could remove guzzle and not worry about that as Guzzle is technically in the oauth lib and additionally the class is only used to catch an exception which if the class doesnt exist PHP doesnt care (for exception catches only). The only issue is that in your IDE (Looks like you use PHPStorm) it would have the dreaded red underline. We could also just catch the base Exception class instead which isn't a problem where it's at

Skullbock commented:
> @Skullbock My only concern was with you exposing credentials you dont want to expose. However if github actions allow you to add credentials without showing the rest of us then it's a moot point :-)

Yes, github allows you to store "secrets" in the repo to use during the execution of actions

> Oh and Guzzle is still a dep because I was catching a guzzle exception of ClientException in ONE place. We could remove guzzle and not worry about that as Guzzle is technically in the oauth lib and additionally the class is only used to catch an exception which if the class doesnt exist PHP doesnt care (for exception catches only). The only issue is that in your IDE (Looks like you use PHPStorm) it would have the dreaded red underline. We could also just catch the base Exception class instead which isn't a problem where it's at

I would remove the dep and catch a generic exception instead, or use the exception without having the dependency, since that way we let the base lib declare the correct guzzle dep versions

tm1000 committed:
remove guzzle since its in oauth2-client, also remove ext-json as its in php 8 and additionally needed by oauth2-client anyways

tm1000 commented:
@Skullbock ok! I removed guzzle (but I kept the exception reference because I need to check the body from guzzle if its a client exception) and also removed ext-json since its required for oauth to even work, additionally in php 8 you can't have php without it

Skullbock commented:
Good! I'll take a look at it next week to see if I can just release it as a minor or not 😊

Thanks again!

tm1000 commented:
@Skullbock were you ever able to look at this?

Skullbock commented:
Hi @tm1000 yes, i would like to release a new major version with this included.
Can you also propose a readme update with the new instructions?

Thanks

tm1000 commented:
@Skullbock as far as I am aware the readme doesnt need to be updated as your wrapper still applies. The only difference is the underlying oauth code is maintained by @thephpleague instead of you

Skullbock commented:
Hi @tm1000 i was thinking of releasing a new major version of this package that allows for this implementation:

- use of one of the two clients listed here: [https://oauth2-client.thephpleague.com/providers/thirdparty/](https://oauth2-client.thephpleague.com/providers/thirdparty/) to use the oauth2 league package
- cleanup the usage of `spatie/enum` to avoid further conflicts
- enable automatic testing again
- support for php 7.3+ and  8.0 and php unit 7/8/9

I will try to work on it in the coming days / weeks so any work on the zoho crm / books stuff gets also easier
Any feedback from you guys is more than welcome

tm1000 commented:
Are you going to have books/crm use on of those two libraries or this use
one of those two libraries?

I think the thing I liked best about yours is the ability to change out
guzzleclient which makes it easy for mocking. I can’t remember if the other
two allow that

On Tue, Feb 2, 2021 at 7:07 AM Daniele Rosario <notifications@github.com>
wrote:

> Hi @tm1000 <https://github.com/tm1000> i was thinking of releasing a new
> major version of this package that allows for this implementation:
>
>    - use of one of the two clients listed here:
>    https://oauth2-client.thephpleague.com/providers/thirdparty/ to use
>    the oauth2 league package
>    - cleanup the usage of spatie/enum to avoid further conflicts
>    - enable automatic testing again
>    - support for php 7.3+ and 8.0 and php unit 7/8/9
>
> I will try to work on it in the coming days / weeks so any work on the
> zoho crm / books stuff gets also easier
> Any feedback from you guys is more than welcome
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Weble/ZohoClient/pull/6#issuecomment-771700402>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEJYIEDJTRO5BO6VOKAP5DS5AILVANCNFSM4TEFYZKA>
> .
>
-- 

Andrew Nagy

Skullbock commented:
Yes, that's the idea

it allows for overriding the default guzzle client (supports 6 and 7)
https://github.com/thephpleague/oauth2-client/blob/66478225f9b1b523f0e5bb49a5d8350a917cad71/src/Provider/AbstractProvider.php#L135

tm1000 commented:
Ha yes I forgot it’s an option for the league.


I originally looked into those two but then didn’t work which is I why I
did the work to get it to work directly against yours.

I’ll admit the only downside to using one of those two is the wasted time
on my part but at least it got the ball rolling

On Tue, Feb 2, 2021 at 8:26 AM Daniele Rosario <notifications@github.com>
wrote:

> Yes, that's the idea
>
> it allows for overriding the default guzzle client (supports 6 and 7)
>
> https://github.com/thephpleague/oauth2-client/blob/66478225f9b1b523f0e5bb49a5d8350a917cad71/src/Provider/AbstractProvider.php#L135
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Weble/ZohoClient/pull/6#issuecomment-771761094>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEJYIEQ3HJ74N2FE4FA3DTS5ARUNANCNFSM4TEFYZKA>
> .
>
-- 

Andrew Nagy

Skullbock commented:
I'm not saying that your implementation is bad mind you! It's great! i'm just trying to lower the amount of code to mantain and leveraging a semi-official client seems the way to go for me

tm1000 commented:
Makes sense

Do you have a choice on which one you want to use? As I am in the middle of
implementing this and books and crm as service providers for my own work.
However if you have one then I can work on implementing your crm and books
with one of those two

Also understood if you haven’t had a chance to look yet. I’ll keep moving
forward on my work and when you move I’ll slowly move mine as well

On Tue, Feb 2, 2021 at 8:38 AM Daniele Rosario <notifications@github.com>
wrote:

> I'm not saying that your implementation is bad mind you! It's great! i'm
> just trying to lower the amount of code to mantain and leveraging a
> semi-official client seems the way to go for me
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Weble/ZohoClient/pull/6#issuecomment-771770929>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEJYIHYFE6CBD3AXB4XB3DS5AS7LANCNFSM4TEFYZKA>
> .
>
-- 

Andrew Nagy

Skullbock commented:
Both seems ok to me and supported at the same level
I would take the latest updated one :)

tm1000 commented:
Something to consider as well. Yours supports caching of tokens. I don’t
think either of those two support that. Which is something I found useful.
Though it could be extracted I suppose

On Tue, Feb 2, 2021 at 8:56 AM Daniele Rosario <notifications@github.com>
wrote:

> Both seems ok to me and supported at the same level
> I would take the latest updated one :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Weble/ZohoClient/pull/6#issuecomment-771786154>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEJYIC5RIF2KJHBNQNQV4LS5AVDPANCNFSM4TEFYZKA>
> .
>
-- 

Andrew Nagy

Skullbock commented:
Will think some more on it for sure :)

On Tue, Feb 2, 2021 at 6:16 PM Andrew Nagy <notifications@github.com> wrote:

> Something to consider as well. Yours supports caching of tokens. I don’t
> think either of those two support that. Which is something I found useful.
> Though it could be extracted I suppose
>
> On Tue, Feb 2, 2021 at 8:56 AM Daniele Rosario <notifications@github.com>
> wrote:
>
> > Both seems ok to me and supported at the same level
> > I would take the latest updated one :)
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/Weble/ZohoClient/pull/6#issuecomment-771786154>, or
> > unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AAEJYIC5RIF2KJHBNQNQV4LS5AVDPANCNFSM4TEFYZKA
> >
> > .
> >
> --
>
> Andrew Nagy
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Weble/ZohoClient/pull/6#issuecomment-771806049>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINRU4X6SQXOYGL6MHE5LDS5AXNBANCNFSM4TEFYZKA>
> .
>


-- 
Daniele Rosario
CTO

Weble.it <http://weble.it/>
Via Enrico Fermi 265
36100 Vicenza
+39 0444 130094 <+390444130094>
+39 328 3017134 <+393283017134>

tm1000 commented:
Replaced by #7
